sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -933,19 +934,34 @@ object DDLUtils {
         case HIVE_PROVIDER =>
           val serde = table.storage.serde
           if (serde == HiveSerDe.sourceToSerDe("orc").get.serde) {
-            OrcFileFormat.checkFieldNames(schema)
+            checkDataColNames("orc", schema)
           } else if (serde == HiveSerDe.sourceToSerDe("parquet").get.serde ||
             serde == Some("parquet.hive.serde.ParquetHiveSerDe") ||
             serde == Some("org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe")) {
-            ParquetSchemaConverter.checkFieldNames(schema)
+            checkDataColNames("parquet", schema)
           }
-        case "parquet" => ParquetSchemaConverter.checkFieldNames(schema)
-        case "orc" => OrcFileFormat.checkFieldNames(schema)
+        case "parquet" => checkDataColNames("parquet", schema)
+        case "orc" => checkDataColNames("orc", schema)
         case _ =>
       }
     }
   }
 
+  def checkDataColNames(provider: String, schema: StructType): Unit = {
+    try {
+      DataSource.lookupDataSource(provider, SQLConf.get).getConstructor().newInstance() match {
+        case f: FileFormat => DataSourceUtils.verifySchema(f, schema)
+        case f: FileDataSourceV2 =>
+          f.getTable(schema, Array.empty, Map.empty[String, String].asJava).schema()
+        case _ =>
+      }
+    } catch {
+      case e: AnalysisException if e.getMessage.contains("contains invalid character") => throw e
```

Comment:
it's a bit hacky to check on the error message here. how about
```
val source = try {
  DataSource.lookupDataSource(provider, SQLConf.get).getConstructor().newInstance()
} catch {
  case NonFatal(e) =>
    logError...
    return
}

source match ...

```

Reply:
Done